src/mod_cache_users.erl

```diff
@@ -0,0 +1,262 @@
+                true -> true
+            end;
+        [] ->
+           is_member(ParentTab, Key, ets:prev(ParentTab, SegmentKey))
+    end.
+
+%% There's a chance that by the time we insert in the ets table, this table is not
+%% the first anymore because the cleaner has taken action and pushed it behind.
+%% That's fine, worst case this record will live a segment less than expected.
+put_member(ParentTab, Key, SegmentKey) ->
+    EtsSegment = ets:lookup_element(ParentTab, SegmentKey, 2),
+    ets:insert(EtsSegment, {Key}).
+
+-spec start_server(mongooseim:host_type(), gen_mod:module_opts()) -> any().
+start_server(HostType, Opts) ->
+    ParentTab = tbl_name(HostType),
+    Spec = #{id => ParentTab, start => {?MODULE, start_link, [ParentTab, Opts]},
+             restart => permanent, shutdown => 5000,
+             type => worker, modules => [?MODULE]},
+    {ok, _} = ejabberd_sup:start_child(Spec).
+
+-spec stop_server(mongooseim:host_type()) -> any().
+stop_server(HostType) ->
+    ok = ejabberd_sup:stop_child(tbl_name(HostType)).
+
+%%--------------------------------------------------------------------
+%% gen_server callbacks
+%%--------------------------------------------------------------------
+
+-record(cache_state, {tab :: atom(),
+                      ttl = timer:hours(8) :: timeout()}).
+
+rotate_and_purge_last_segment(#cache_state{tab = ParentTab}) ->
+    First = ets:first(ParentTab),
+    Last = ets:last(ParentTab),
+    EtsSegment = case ets:info(ParentTab, size) of
+                     1 ->
+                         ets:lookup_element(ParentTab, First, 2);
+                     _ ->
+                         [{_, E}] = ets:take(ParentTab, First),
+                         true = ets:insert_new(ParentTab, {Last + 1, E}),
+                         E
+                 end,
+    ets:delete_all_objects(EtsSegment),
+    ok.
+
+iterate_through_tables(ParentTab, Action) when is_atom(ParentTab), is_function(Action, 1) ->
+    iterate_(ParentTab, Action, ets:first(ParentTab)).
+
+iterate_(_, _, '$end_of_table') ->
+    ok;
+iterate_(ParentTab, Action, SegmentKey) when is_atom(ParentTab), is_function(Action, 1) ->
+    case ets:lookup(ParentTab, SegmentKey) of
+        [{_, EtsSegment}] ->
+            Action(EtsSegment),
+            iterate_(ParentTab, Action, ets:next(ParentTab, SegmentKey));
+        [] ->
+            iterate_(ParentTab, Action, ets:next(ParentTab, SegmentKey))
+    end.
+
+-spec start_link(atom(), gen_mod:module_opts()) -> {ok, pid()}.
+start_link(ParentTab, Opts) ->
+    gen_server:start_link({local, ParentTab}, ?MODULE, [ParentTab, Opts], []).
+
+init([ParentName, Opts]) ->
+    TTL0 = gen_mod:get_opt(ttl, Opts, 8 * 60), %% 8h
+    N    = gen_mod:get_opt(number_of_segments, Opts, 3),
+    EtsOpts = [ordered_set, named_table, protected, {read_concurrency, true}],
+    ets:new(ParentName, EtsOpts),
+    lists:foreach(
+      fun(I) ->
+              %% Note, these names are only for debugging purposes,
```

Comment:
What are "these names"? The tables are not named.

Reply:
Ah, they were named before and then when un-naming I forgot to remove the comment